### ISSUE TITLE
Add support for phased data (and return extra variant information)

### ISSUE THREAD

**Issue by hyanwong** [limix/bgen-reader-py#8]
It's not clear to me how to read extra variant information - for instance I'm trying to check on the `phased` flag as described in "Layout 2 probability data storage" on the BGEN definition page (http://www.well.ox.ac.uk/~gav/bgen_format/). How can I access this using bgen-reader-py ?

**Comment by horta** [comment]
Actually the reader right now does not support phased genotype. It will fail if you try to read phased genotypes. I will label this thread as an "enchancement" and see if I can implement this feature this weekend. Thanks for the feedback!

**Comment by hyanwong** [comment]
Ah, thanks. That would be really useful. But I should say that I have tried on the UK-BB phased haplotype data just now, and I can successfully extract variants, samples, and genotypes arrays:


```
>>> bgen = read_bgen("ukb_hap_chr20_v2.bgen", verbose=False)
>>> print(bgen['variants'])
       id           rsid chrom       pos  nalleles allele_ids
0      20      rs6076506           63231         2        T,G
...
```

 The genotype array looks something like
```
>>> print(bgen['genotype'][X].compute())
[[ 1.  1. -1.]
 [ 1.  1. -1.]
...
```

**Comment by horta** [comment]
hum i guess the C library is not applying the assertion (probably because it was compiled for maximum performance). well, i will have a look and provide the information about the phase anyway =)

**Comment by hyanwong** [comment]
That would be really useful, thanks. It would be incredible if I could read this stuff next week. NB: the other python bgen library (https://lemieuxl.github.io/pybgen/) says

`ValueError: haplotypes/ukb_hap_chr20_v2.bgen: only accepting unphased data`

So if you do implement it, yours will be the only python library I can find that does so.

**Comment by hyanwong** [comment]
p.s. If I know that this data phased, is the `bgen['genotype']` data usable/meaningful at all?

**Comment by horta** [comment]
Hum, I'm not sure. I don't remember the requirements for phased.

**Comment by hyanwong** [comment]
OK, thanks. I've never used BGEN until now, so I've even more in the dark than you.

**Comment by hyanwong** [comment]
How's the phased data readability going? If it helps, the genotypes reported by `bgen['genotype'][X].compute()` from a phased version of UK-BB all fall into one of the 4 following tuple types, which I suspect may correspond to the phased genotypes AA, AB, BA, and BB, for the 2 alleles A and B:

`{(1.0, 0.0, 0.0), (0.0, 1.0, 0.0), (1.0, 1.0, -1.0), (0.0, 0.0, 1.0)}`

Let me know if I can help...

**Comment by horta** [comment]
Hi @hyanwong , didn't have time yet to work on that, sorry. If you now C and BGEN format, you might be able to help me by understanding this: https://github.com/limix/bgen/blob/master/src/layout/two.c#L40

But if not, that is fine too. The code as a bit complicated, as the BEN format is not trivial.

**Comment by hyanwong** [comment]
I'm actually chatting to Gavin Band, author of the bgen format, now. He thinks that for phased data, the 3rd number for each bgen['genotype'][X].compute() result you return is redundant. The first 2 numbers give the probability of the 1st allele in the (e.g. maternal) haplotype, and the probability of the 1st allele in the (e.g. paternal) haplotype. The probability of the 2nd allele in each case is just one minus that number. So the 3rd number reported by you library is redundant.

I will check this with the rbgen library that he uses.

**Comment by hyanwong** [comment]
I have confirmed this by comparison with the rbgen library. When I do something in rbgen like `D = bgen.load('phased_data.bgn',  max_entries_per_sample=4); D$data[var_id,X:X+10,1:4]` I get a 4 column array like this

```
g=0 g=1 g=2 g=3
  0   1   1   0
  0   1   1   0
  1   0   0   1
  1   0   0   1
  1   0   0   1
  0   1   1   0
  1   0   0   1
  1   0   0   1
  0   1   0   1
  0   1   1   0
```

 where `g=1` = 1-`g=0` and `g=3` = 1-`g=2`, and where according to the spec, `g=0` is the freq in one haplotype (e.g. maternal) and `g=2` is the freq in the other (e.g. paternal). This corresponds to the equivalent read-out from your python library:

```
>>> bgen['genotype'][var_id-1,X-1:X-1+10:,0:2].compute()
array([[0., 1.],
       [0., 1.],
       [1., 0.],
       [1., 0.],
       [1., 0.],
       [0., 1.],
       [1., 0.],
       [1., 0.],
       [0., 0.],
       [0., 1.]])

```

So the phased data is there and interpretable in the array returned by your bgen library.

**Comment by horta** [comment]
Hi @hyanwong , thanks a lot for the feedback! I have implemented a draft for reading phased data. Is it feasible for you to send me an example of phased bgen file? For testing purposes as I dont have any...

**Comment by hyanwong** [comment]
I can't really send any from the datasets I'm using, as these are restricted access. Will see if Gavin has a test set.

**Comment by hyanwong** [comment]
Gavin helpfully pointed me to 

https://bitbucket.org/gavinband/bgen/src/default/example/haplotypes.bgen

I guess that's enough?

**Comment by horta** [comment]
It should be. Thanks! I will tests it here.

**Comment by horta** [comment]
Hi @hyanwong , could you help me understand the following notation?
In the url you showed me there is also a text file that explains the haplotypes.bgen:

```
1 SNP1 RS1 1 A G | 0 0 | 1 0 | 0 1 | 1 1
1 SNP2 RS2 2 A G | 1 0 | 0 1 | 1 1 | 0 0
1 SNP3 RS3 3 A G | 0 1 | 1 1 | 0 0 | 1 0
1 SNP4 RS4 4 A G | 1 1 | 0 0 | 1 0 | 0 1
```

(I've added | to separate the individuals).
SNP1, first individual, has 0 0. For me, it would imply that allele A of SNP1 has 0% of probability of pertaining to the first chromossome, and 100% of pertaining to the second. The same with the second allele: 0% of probability of pertaining to the first chromossome and 100% of pertaining to the second chromossome. But when I read that file using my updated reader, it shows everything switched:

```
 1.0 0.0 1.0 0.0 |  0.0 1.0 1.0 0.0 |  1.0 0.0 0.0 1.0 |  0.0 1.0 0.0 1.0 |
 0.0 1.0 1.0 0.0 |  1.0 0.0 0.0 1.0 |  0.0 1.0 0.0 1.0 |  1.0 0.0 1.0 0.0 |
 1.0 0.0 0.0 1.0 |  0.0 1.0 0.0 1.0 |  1.0 0.0 1.0 0.0 |  0.0 1.0 1.0 0.0 |
 0.0 1.0 0.0 1.0 |  1.0 0.0 1.0 0.0 |  0.0 1.0 1.0 0.0 |  1.0 0.0 0.0 1.0 |
```

The first line, of the first column (1.0 0.0 1.0 0.0) means that the first allele of SNP1 has 100% of chance of pertaining to the first chromossome, and 0% to the second. And the same for the second allele.

It happens in every case. Do you know what I'am missing?

**Comment by hyanwong** [comment]
That's the same result as output by rbgen. So I think it's some confusion about what the meaning of 0 and 1 is. In particular, I think the probability for alleles is taken from the *second* and *fourth* items in (1.0 0.0 1.0 0.0), but will check and get back to you.

**Comment by gavinband** [comment]
```
 1.0 0.0 1.0 0.0 |  0.0 1.0 1.0 0.0 |  1.0 0.0 0.0 1.0 |  0.0 1.0 0.0 1.0 |
 0.0 1.0 1.0 0.0 |  1.0 0.0 0.0 1.0 |  0.0 1.0 0.0 1.0 |  1.0 0.0 1.0 0.0 |
 1.0 0.0 0.0 1.0 |  0.0 1.0 0.0 1.0 |  1.0 0.0 1.0 0.0 |  0.0 1.0 1.0 0.0 |
 0.0 1.0 0.0 1.0 |  1.0 0.0 1.0 0.0 |  0.0 1.0 1.0 0.0 |  1.0 0.0 0.0 1.0 |
```

> The first line, of the first column (1.0 0.0 1.0 0.0) means that the first allele of SNP1 has 100% of chance of pertaining to the first chromossome, and 0% to the second. And the same for the second allele.

No, the first entry **1.0** 0.0 **1.0** 0.0 means that chromosome 1 carries allele 1 with probability 1 (and allele2 with probabiltiy 0), and likewise for chromosome 2.  I have marked in bold the values that are actually stored in the file: they are the probability values for allele1 for the two chromosomes.  The text file haplotypes.hap instead lists probabilities of allele 2, as in SHAPEIT format or IMPUTE allele probs format.

In other words all this looks ok to me.

Hope this helps,
g.

**Comment by hyanwong** [comment]
Thanks Gavin. Sorry I got this the wrong way about.

**Comment by horta** [comment]
Thanks both. Now I get it! I will release the bgen-reader (and the necesary bgen C library) tomorrow with phased alleles!

**Comment by hyanwong** [comment]
Great, thanks for your quick work. Let me know when it is ready for testing.

Yan

**Comment by horta** [comment]
Hi @hyanwong , I've finished the phased implementation, among other things: conda update -c conda-forge bgen-reader . Let me know how it goes =)

**Comment by hyanwong** [comment]
Looks great, thanks for your hard work. So this is bgen_reader v 2.0.0 now?

**Comment by horta** [comment]
Yep =)

**Comment by hyanwong** [comment]
After `conda update -c conda-forge bgen-reader` I'm getting:

```
    from ._ffi import ffi as _
ImportError: libbgen.so.2.0.0: cannot open shared object file: No such file or directory
It is likely caused by a broken installation of this package.
Please, make sure you have a C compiler and try to uninstall
and reinstall the package again.
```

Is there a missing dep?

**Comment by hyanwong** [comment]
Doh, sorry. Guess I should have done `conda install -f -c conda-forge bgen` beforehand. Will do so now

**Comment by horta** [comment]
Actually that is a bug. Will fix it, thanks =)

**Comment by hyanwong** [comment]
I now get this reading error

```
>>> bgen = read_bgen(bgen_file, verbose=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "miniconda3/lib/python3.6/site-packages/bgen_reader/_reader.py", line 260, in read_bgen
    G, X = _read_genotype(index, nsamples, nvariants, nalls, size, verbose)
  File "miniconda3/lib/python3.6/site-packages/bgen_reader/_reader.py", line 196, in _read_genotype
    return da.concatenate(genotype), da.concatenate(X)
  File "miniconda3/lib/python3.6/site-packages/dask/array/core.py", line 2607, in concatenate
    % str(seq[0].shape))
ValueError: Tried to concatenate arrays with unknown shape (13, 487409, nan).  To force concatenation pass allow_unknown_chunksizes=True.
```

**Comment by horta** [comment]
hmm which version of dask you have?
```
import dask
print(dask.__version__)
```

I think the latest version of dask won't give that error...

**Comment by hyanwong** [comment]
```
>>> import dask
>>> print(dask.__version__)
0.17.5
```

**Comment by horta** [comment]
hum I haven't seen that error because the examples have a single chunk :/ Fixing that.

**Comment by hyanwong** [comment]
Right - let me know when I can download (e.g. from git) and test. Also let me know if you would prefer me to report things like this in this issue, or in create a new issue for each.

**Comment by horta** [comment]
It is fine here. I just released 2.0.1 on pip: pip install bgen-reader --upgrade
It should fix it!

**Comment by hyanwong** [comment]
Seems to work now, thanks. I should note that doing `pip install bgen-reader --upgrade` in my installation complains about cftime, so I wonder if this (cython) is another dependency that needs including. The installation completes, however.

```
pip install bgen-reader --upgrade   
Collecting bgen-reader
  Downloading ...
Building wheels for collected packages: bgen-reader
  Running setup.py bdist_wheel for bgen-reader ... done
cftime 1.0.0 requires cython, which is not installed.
...
Successfully installed bgen-reader-2.0.1
```

**Comment by hyanwong** [comment]
Oh, spoke too soon:

```
bgen = bgen_reader.read_bgen(bgen_file, verbose=False) #now fine in 2.0.1
a = bgen['genotype'][1,:,0:2].compute() #used to work
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "miniconda3/lib/python3.6/site-packages/dask/array/core.py", line 1317, in __getitem__
    dsk, chunks = slice_array(out, self.name, self.chunks, index2)
  File "miniconda3/lib/python3.6/site-packages/dask/array/slicing.py", line 157, in slice_array
    dsk_out, bd_out = slice_with_newaxes(out_name, in_name, blockdims, index)
  File "miniconda3/lib/python3.6/site-packages/dask/array/slicing.py", line 179, in slice_with_newaxes
    dsk, blockdims2 = slice_wrap_lists(out_name, in_name, blockdims, index2)
  File "miniconda3/lib/python3.6/site-packages/dask/array/slicing.py", line 234, in slice_wrap_lists
    return slice_slices_and_integers(out_name, in_name, blockdims, index)
  File "miniconda3/lib/python3.6/site-packages/dask/array/slicing.py", line 279, in slice_slices_and_integers
    raise ValueError("Arrays chunk sizes are unknown: %s", shape)
ValueError: ('Arrays chunk sizes are unknown: %s', (17197, 487409, nan))

```

which is with the latest code:

```
>>> bgen_reader.__version__
'2.0.1'
```

**Comment by horta** [comment]
I think that you will receive that warning if you try to install any package. Thanks!

**Comment by hyanwong** [comment]
[You mean the "requires cython" one]. I'm still hitting the problem of indexing with e.g. `bgen['genotype'][1,:,0:2].compute()` giving `ValueError: ('Arrays chunk sizes are unknown: %s', (17197, 487409, nan))`

**Comment by hyanwong** [comment]
Actually, 2 errors. One with slicing and one with using an array for indexes:

```
>>> bgen['genotype'][1,:,0:2]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "miniconda3/lib/python3.6/site-packages/dask/array/core.py", line 1317, in __getitem__
    dsk, chunks = slice_array(out, self.name, self.chunks, index2)
  File "miniconda3/lib/python3.6/site-packages/dask/array/slicing.py", line 157, in slice_array
    dsk_out, bd_out = slice_with_newaxes(out_name, in_name, blockdims, index)
  File "miniconda3/lib/python3.6/site-packages/dask/array/slicing.py", line 179, in slice_with_newaxes
    dsk, blockdims2 = slice_wrap_lists(out_name, in_name, blockdims, index2)
  File "miniconda3/lib/python3.6/site-packages/dask/array/slicing.py", line 234, in slice_wrap_lists
    return slice_slices_and_integers(out_name, in_name, blockdims, index)
  File "miniconda3/lib/python3.6/site-packages/dask/array/slicing.py", line 279, in slice_slices_and_integers
    raise ValueError("Arrays chunk sizes are unknown: %s", shape)
ValueError: ('Arrays chunk sizes are unknown: %s', (19139, 487409, nan))
```

and

```
>>> bgen['genotype'][1,:,[0,2]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "miniconda3/lib/python3.6/site-packages/dask/array/core.py", line 1321, in __getitem__
    return Array(dsk2, out, chunks, dtype=self.dtype)
  File "miniconda3/lib/python3.6/site-packages/dask/array/core.py", line 1037, in __new__
    self._chunks = normalize_chunks(chunks, shape)
  File "miniconda3/lib/python3.6/site-packages/dask/array/core.py", line 1932, in normalize_chunks
    raise ValueError("Empty tuples are not allowed in chunks. Express "
ValueError: Empty tuples are not allowed in chunks. Express zero length dimensions with 0(s) in chunks
```

**Comment by horta** [comment]
I mean the "requires cython" yes.

That error comes from dask and is not very helpful. But you can avoid that by doing: `bgen['genotype'][1,:,:].compute()`. I had to adopt dask with unknown dimension (the last one) because bgen allows variants/samples with different ploidy/phase/number of alleles. Sorry for that. I hope dask package come up with a better solution for that case in the future...

**Comment by hyanwong** [comment]
Can you code it so as to pass in the dimension when doing `read_bgen`. That's what the R library does, I think (e.g. ` max_entries_per_sample=4`)?

I guess the alternative is to retrieve everything and subset afterwards, e.g. `bgen['genotype'][1,:,:].compute()[:,[0,2]]`. Presumably this uses up double the amount of resident memory than indexing before the compute() call?

**Comment by horta** [comment]
`bgeno['genotype'][1,:,[0,2]].compute()` and `bgeno['genotype'][1,:,:].compute()[:,[0,2]]` will use the same amount of memory. I would not worry about the last dimension. What might cause problem memory-wise is the first dimension, but that problem is the same for the old or new version of bgen-reader.

I implemented bgen-reader in a way to avoid increasing the interface complexity, for example by not adding `max_entries_per_sample` and relying on dask to reshape accordingly. I think this approach might get easier over time as dask is a young library but the deveopers are working hard on it.

**Comment by hyanwong** [comment]
Actually in my case the 2nd dimension is *way* larger than the first (array is roughly of dimensions [16000, 500000,4]). But point taken, thanks. I'm probably done here.
